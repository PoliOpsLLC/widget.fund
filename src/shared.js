import fetch from 'unfetch';

export const bootstrap = (key, errorMessage = 'something went wrong') => {
    const handleError = err => {
        const error = document.createElement('span');
        error.innerHTML = errorMessage;
        document.body.appendChild(error);
        console.log(err);
    };

    const options = {
        method: 'POST',
        credentials: 'include',
        headers: {
            'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8',
        },
        body: packParams({ key }),
    };

    if (process.env.CREDENTIALS) {
        options.headers.Authorization = `Basic ${btoa(process.env.CREDENTIALS)}`;
    }

    console.log(options);

    return fetch(`${process.env.API_URL}${process.env.BOOTSTRAP_ENDPOINT}`, options)
        .then(resp => {
            if (!resp.ok) throw Error(resp.statusText);
            return resp.json();
        })
        .catch(handleError);
};

export const flushPromises = () => new Promise(resolve => setImmediate(resolve));

export const packParams = params => {
    return Object.keys(params).map(key => {
        return `${encodeURIComponent(key)}=${encodeURIComponent(params[key])}`;
    }).join('&');
};

export const simulateOn = (target, name, props = {}) => {
    const node = document.querySelector(target);
    const event = new Event(name, { bubbles: true });
    Object.assign(event, props);
    node.dispatchEvent(event);
};

export default { bootstrap, flushPromises, packParams, simulateOn };
