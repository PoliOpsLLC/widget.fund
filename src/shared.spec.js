import nock from 'nock';

import { bootstrap, debounce, packParams } from './shared';

jest.useFakeTimers();

describe('Shared', () => {
    describe('bootstrap', () => {
        it('handles network errors', async () => {
            nock(process.env.API_URL)
                .post(process.env.BOOTSTRAP_ENDPOINT, 'key=1234')
                .replyWithError('oops!');
            await bootstrap(1234);
            expect(document.querySelector('span')).toMatchSnapshot();
        });

        it('returns data on success', async () => {
            nock(process.env.API_URL)
                .post(process.env.BOOTSTRAP_ENDPOINT, 'key=1234')
                .reply(200, { token: 5678 });
            const resp = await bootstrap(1234);
            expect(resp.token).toEqual(5678);
        });
    });

    it('debounce', () => {
        const fn = jest.fn();
        const debounced = debounce(fn, 500);
        debounced();
        debounced();
        debounced();
        expect(fn).not.toBeCalled();
        jest.advanceTimersByTime(499);
        expect(fn).not.toBeCalled();
        jest.advanceTimersByTime(1);
        expect(fn).toHaveBeenCalledTimes(1);
    });

    it('packParams', () => {
        expect(packParams({
            reserved: ';,/?:@&=+$',
            unescaped: '-_.!~*\'()',
            alphanum: 'ABC abc #123',
            ignored: undefined,
            stripped: null,
            missing: '',
            disregarded: 0,
        })).toMatchSnapshot();
    });
});
