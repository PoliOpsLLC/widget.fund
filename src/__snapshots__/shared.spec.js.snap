// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shared bootstrap handles network errors 1`] = `
<span>
  oops!
</span>
`;

exports[`Shared packParams 1`] = `"reserved=%3B%2C%2F%3F%3A%40%26%3D%2B%24&unescaped=-_.!~*'()&alphanum=ABC%20abc%20%23123"`;
